build/circle-test.sh

```diff
@@ -1,105 +1,93 @@
 #!/bin/bash
 
-set -e
+set -euo pipefail
 
-outdir="${TMPDIR}"
-if [ "${CIRCLE_ARTIFACTS}" != "" ]; then
+if [ -n "${CIRCLE_ARTIFACTS-}" ]; then
   outdir="${CIRCLE_ARTIFACTS}"
+elif [ -n "${TMPDIR-}" ]; then
+  outdir="${TMPDIR}"
+else
+  outdir=""
 fi
 
 builder=$(dirname $0)/builder.sh
+match='^panic|^[Gg]oroutine \d+|(read|write) by.*goroutine|DATA RACE'
+
+open_github_issue() {
+  if [ "${CIRCLE_BRANCH-}" = "master" ] && [ -n "${GITHUB_API_TOKEN-}" ]; then
+    curl -X POST -H "Authorization: token ${GITHUB_API_TOKEN}" \
+      "https://api.github.com/repos/${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}/issues" \
+      -d "{ \"title\": \"Test failure in CI build ${CIRCLE_BUILD_NUM}\", \"body\": \"The following test appears to have failed:\n\n[#${CIRCLE_BUILD_NUM}](https://circleci.com/gh/${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}/${CIRCLE_BUILD_NUM}):\n\n\`\`\`\n$(python -c 'import json,sys; print json.dumps(sys.stdin.read()).strip("\"")' < ${outdir}/excerpt.txt)\n\`\`\`\nPlease assign, take a look and update the issue accordingly.\", \"labels\": [\"test-failure\"] }" > /dev/null
+    echo "Found test/race failures in test logs, see excerpt.log and the newly created issue on our issue tracker"
+  fi
+}
+
+prepare_artifacts() {
+  # Translate the log output to xml to integrate with CircleCI
+  # better.
+  if [ -n "${CIRCLE_TEST_REPORTS-}" ]; then
+    if [ -f "${outdir}/test.log" ]; then
+      mkdir -p "${CIRCLE_TEST_REPORTS}/test"
+      ${builder} go2xunit < "${outdir}/test.log" \
+        > "${CIRCLE_TEST_REPORTS}/test/test.xml"
+    fi
+    if [ -f "${outdir}/testrace.log" ]; then
+      mkdir -p "${CIRCLE_TEST_REPORTS}/race"
+      ${builder} go2xunit < "${outdir}/testrace.log" \
+        > "${CIRCLE_TEST_REPORTS}/race/testrace.xml"
+    fi
+  fi
+
+  # Generate the slow test output.
+  find "${outdir}" -name 'test*.log' -type f -exec \
+    grep -F ': Test' {} ';' | \
+    sed -E 's/(--- PASS: |\(|\))//g' | \
+    awk '{ print $2, $1 }' | sort -n | tail -n 10 \
```

Comment:
Comment that the use of `tail` (instead of a reverse sort with `head`) is to work around the fact that `set -o pipefail` doesn't like it if `sort` exits with a `SIGPIPE` error code.